phoenicis-repository/src/main/java/org/phoenicis/repository/types/GitRepository.java

```diff
@@ -65,72 +69,103 @@ private File createRepositoryLocation(String cacheDirectoryPath) {
         return new File(cacheDirectoryPath + "/git" + hashcode);
     }
 
-    @Override
-    public synchronized RepositoryDTO fetchInstallableApplications() {
-        RepositoryDTO result = null;
-        try {
-            mutex.acquire();
+    private void cloneOrUpdate() throws RepositoryException {
+        synchronized (mutex) {
             try {
                 LOGGER.info("Begin fetching process of " + this);
 
-                boolean folderExists = this.localFolder.exists();
+                boolean lockFileExists = this.lockFile.exists();
 
-                // check that the repository folder exists
-                if (!folderExists) {
-                    LOGGER.info("Creating local folder for " + this);
+                // check that the repository lock file exists
+                if (!lockFileExists) {
+                    LOGGER.info("Creating lock file for " + this);
 
-                    if (!this.localFolder.mkdirs()) {
-                        throw new RepositoryException("Couldn't create local folder for " + this);
+                    try {
+                        this.lockFile.getParentFile().mkdirs();
+                        this.lockFile.createNewFile();
```

Comment:
This (i.e. persistent lock management) is not the responsibility of GitRepository.
You should create another class

Reply:
How should that class look like? I think it would have to know internal information of `GitRepository`.

Reply:
What information would it need?

Reply:
E.g. when the lock is required and for which file. I could create a class which locks a resource on request but that would give pretty much the same code as I have now (because that's exactly what the Java API does already).

Reply:
So basically, you need some kind of Map associating a file path with a semaphore, right?

Reply:
I don't think so. But not sure I understood you correctly.